Add a new profile constraint for subject name

fmarco76 opened the pull request:
Implement a new profile constraint, named `RAClientAuthSubjectNameContraint`, to limit the subject name of enrolled certificate when submitted by an agent.

The new constraint get the information for RAClientAuthInput and try to match the client certificate subject with the one in the CSR or, if the client was not using TLS authentication, match the client name.

The new constraint implement similar EST policy on the CA.

ladycfu commented:
@fmarco76 the commit comment says "... to limit the subject name of enrolled certificate when submitted by **an agent**."  I think you meant "...to limit the subject name of enrolled certificate when submitted by **a non-agent user.**"

ladycfu reviewed (approved):
LGTM

fmarco76 commented:
> @fmarco76 the commit comment says "... to limit the subject name of enrolled certificate when submitted by **an agent**." I think you meant "...to limit the subject name of enrolled certificate when submitted by **a non-agent user.**"

Since we are modifying the constraint in the CA when the RA submit a certificate for the client. the submitter is the one sending the request to the CA or the one submitting the request to the RA. I have no particular preference, important to match with documentation.

edewata commented:
Looks like my last PR caused the IPA clone test to fail, so just ignore it, I'll fix it separately.

edewata commented on .github/workflows/est-ds-realm-test.yml:
```diff
@@ -392,6 +392,32 @@ jobs:
+
+      - name: Modify CA Subject Name policy
+        run: |
+          docker exec pki sed -i 's/policyset.serverCertSet.list=1,2,3,4,6,7,8,10,20,22/policyset.serverCertSet.list=1,2,3,4,6,7,8,10,21,22/g' \
+              /etc/pki/pki-tomcat/ca/profiles/ca/estServiceCert.cfg
```
Could you add a comment in the test describing what's being changed?

edewata commented on .github/workflows/est-ds-realm-test.yml:
```diff
@@ -392,6 +392,32 @@ jobs:
+          docker exec -e EST_OPENSSL_CACERT=cacert.pem pki estclient -e -s pki.example.com -p 8443 \
+              --common-name test-new.example.com -o new_certs3 -u est-test-user -h Secret.123 | tee output || true
+
+          # Request should fails with bad request error (status code 404)
+          STATUS=$(cat output  | sed -nE 's/.* HTTP response from EST server was (.*)$/\1/p')
+          [ "$STATUS" == "BAD REQUEST" ]
```
Should we check for a specific return code (e.g. 401 or 403) or just any 4xx code (bad request) would be accepted? Not sure if 404 (not found) is the proper code for this scenario. Same thing with the next test below.

fmarco76 replied:
The current return code here is 400.  The CA endpoint is identified and process the request but the output is not good. In the 4xx error options I think this is acceptable in this case, the other are more specific but for different cases.

The status is printed by `estclient` and it writes the error code number for some errors while write the error name for other so we cannot get the number in this case.

edewata replied:
OK, it's just the comment says `status code 404`, that probably should be updated.

edewata commented on base/ca/shared/profiles/ca/estServiceCert.cfg:
```diff
@@ -124,24 +124,38 @@ policyset.serverCertSet.8.constraint.params.signingAlgsAllowed=SHA256withRSA,SHA
 policyset.serverCertSet.20.constraint.name=Subject Name Constraint
 # If the subject names in your EST application follow a different pattern,
 # update this parameter to match your application.
-policyset.serverCertSet.20.constraint.params.pattern=CN=[^,]+
-policyset.serverCertSet.20.constraint.params.accept=true
+policyset.serverCertSet.20.constraint.params.pattern=CN=$ra_client_name$
 policyset.serverCertSet.20.default.class_id=subjectNameDefaultImpl
 policyset.serverCertSet.20.default.name=Subject Name Default
 policyset.serverCertSet.20.default.params.name=CN=$request.req_subject_name.cn$
-# This component (`21`) copies the Common Name value to the Subject
+# 
+policyset.serverCertSet.21.constraint.class_id=subjectNameConstraintImpl
+policyset.serverCertSet.21.constraint.name=Subject Name Constraint
+# If the subject names in your EST application follow a different pattern,
+# update this parameter to match your application.
+policyset.serverCertSet.21.constraint.params.pattern=CN=[^,]+
+policyset.serverCertSet.21.constraint.params.accept=true
+policyset.serverCertSet.21.default.class_id=subjectNameDefaultImpl
+policyset.serverCertSet.21.default.name=Subject Name Default
+policyset.serverCertSet.21.default.params.name=CN=$request.req_subject_name.cn$
+# This component (`22`) copies the Common Name value to the Subject
 # Alternative Name extension (as a dNSName value) if and only if it
 # has the appearance of a DNS name (with more than one label/component).
 #
-# To disable this extension remove `21` from the
+# To disable this extension remove `22` from the
 # `policyset.serverCertSet.list` parameter.
-policyset.serverCertSet.21.constraint.class_id=noConstraintImpl
-policyset.serverCertSet.21.constraint.name=No Constraint
-policyset.serverCertSet.21.default.class_id=commonNameToSANDefaultImpl
-policyset.serverCertSet.21.default.name=CN To SAN Default
-policyset.serverCertSet.list=1,2,3,4,6,7,8,10,20,21
+policyset.serverCertSet.22.constraint.class_id=noConstraintImpl
+policyset.serverCertSet.22.constraint.name=No Constraint
+policyset.serverCertSet.22.default.class_id=commonNameToSANDefaultImpl
+policyset.serverCertSet.22.default.name=CN To SAN Default
+policyset.serverCertSet.list=1,2,3,4,6,7,8,10,20,22
```
Do we need an upgrade script for these changes?

fmarco76 replied:
I am not sure here. The current instance should work and have the same behaviour after the upgrade. If they have some customisation in place the upgrade script could create more problems than solution. We can do some tests later for the upgrade to verify if it is needed.

edewata replied:
I'll let you and @ladycfu decide about this. My concern is that if someone installs EST with PKI 11.6 then later upgrade to PKI 11.7 they won't automatically get the new behavior introduced in PKI 11.7 unless they manually update their config files. Is that OK? Maybe the changes should be documented so people can update their config files manually.

fmarco76 replied:
Yes, when these changes are completed the documentation will be updated. We can explain what and if should be modified.

edewata reviewed (commented):
I have no objection against this PR, just comments/questions, but would it make sense for EST to enforce the policy before sending the request to the CA? The ACME is basically doing that, i.e. validating that the client has the ownership of the DNS name. I'll defer to @ladycfu for the final approval.

fmarco76 commented:
> would it make sense for EST to enforce the policy before sending the request to the CA?

Currently, EST enforce the policy before submitting to the CA and this PR implement exactly the same policy we have. 
When both are enabled they are redundant, and this is the reason I do not think the upgrade script is needed at the moment. However, these policies will be extended to support more complex scenarios with different user roles and the CA has already the capability for that.

edewata commented on tests/ansible/est/tasks/main.yml:
```diff
@@ -230,7 +230,8 @@
       EST_OPENSSL_CACERT: cacert.pem
   loop:
-    - estclient -e -s {{ pki_hostname }} -p 8443 --common-name alice -o . -u alice -h 4me2Test
+# PkiInMemoryRealm has the full name hardcoded and it is "Adminitrsator"
```
typo in comment: `Adminitrsator`

edewata commented:
Thanks for the update! My suggestion is to add either a manual upgrade doc (so the admin can decide whether to upgrade) or an automated upgrade script (that will force the upgrade). That can be added separately. Feel free to merge if @ladycfu is OK.

ladycfu commented:
> > @fmarco76 the commit comment says "... to limit the subject name of enrolled certificate when submitted by **an agent**." I think you meant "...to limit the subject name of enrolled certificate when submitted by **a non-agent user.**"
> 
> Since we are modifying the constraint in the CA when the RA submit a certificate for the client. the submitter is the one sending the request to the CA or the one submitting the request to the RA. I have no particular preference, important to match with documentation.

@fmarco76 The way I'd document it, and the way I'd expect the code to work, would be that the est_ca_user (as per defined in the est pkispawn instructions)  is considered a non-user entity that is a means to provide secure communication between EST and CA.  It serves the same purpose as the "subsystem cert" used by other pk subsystems, which is the TLS client cert used to talk to each other with.  I'd probably refrain from calling it an "agent" to avoid confusion.  On the other hand, the actual user identity that authenticates with EST is to be considered the actual user.  If the user that identifies with EST, and further passed from EST to CA as the authenticated client, is verified by the CA to be a CA agent, then this client is authorized to submit a CSR bearing someone else's subject name and related info, otherwise, it is subject to the new constraint verification where the authenticated id and the CSR SN/SAN must match.

fmarco76 commented:
> > > @fmarco76 the commit comment says "... to limit the subject name of enrolled certificate when submitted by **an agent**." I think you meant "...to limit the subject name of enrolled certificate when submitted by **a non-agent user.**"
> > 
> > 
> > Since we are modifying the constraint in the CA when the RA submit a certificate for the client. the submitter is the one sending the request to the CA or the one submitting the request to the RA. I have no particular preference, important to match with documentation.
> 
> @fmarco76 The way I'd document it, and the way I'd expect the code to work, would be that the est_ca_user (as per defined in the est pkispawn instructions) is considered a non-user entity that is a means to provide secure communication between EST and CA. It serves the same purpose as the "subsystem cert" used by other pki subsystems, which is the TLS client cert used to talk to each other with. I'd probably refrain from calling it an "agent" to avoid confusion. On the other hand, the actual user identity that authenticates with EST is to be considered the actual user. If the user that identifies with EST, and further passed from EST to CA as the authenticated client, is verified by the CA to be a CA agent, then this client is authorized to submit a CSR bearing someone else's subject name and related info, otherwise, it is subject to the new constraint verification where the authenticated id and the CSR SN/SAN must match.

@cfu I have removed the _agent_ from the commit and added a new one checking if the client certificate is owned by a CA agent and if it is the case there is no subject enforcement. I have also added a test.

I think we have to modify the EST user in the CA to be a Subsystem user but this requires change in the installatio

ladycfu commented on .github/workflows/est-ds-realm-test.yml:
```diff
@@ -393,17 +393,39 @@ jobs:
         run: |
```
Could you explain the comment regarding "... and check disabled"?  What "check" was disabled?  And if the "check" was disabled, why expect BAD request?

fmarco76 replied:
Currently, subject check is in EST module and in CA profile. In the previous check the request with different subject was reject from EST subsystem. The above step remove the check from EST subsystem so the subject check is done only by the CA.

In theory, the EST subsystem could support multiple CA/profiles so it should be OK to decide where the subject should be verified.

ladycfu replied:
Ok, the comment might want to explain it more clearly.  Also, to avoid confusion, the original check in EST should be disabled by default so perhaps tests involving checks in the est module could be disabled once this ticket is done.

ladycfu commented on .github/workflows/est-ds-realm-test.yml:
```diff
@@ -393,17 +393,39 @@ jobs:
+          STATUS=$(cat output  | sed -nE 's/.* HTTP response from EST server was (.*)$/\1/p')
+          [ "$STATUS" == "BAD REQUEST" ]
 
       - name: Enroll new certificate with certificate but using different subject and check disabled
```
same question here about "check disabled".

fmarco76 replied:
As above. I could add something to be more explicit

fmarco76 commented on .github/workflows/est-ds-realm-separate-test.yml:
```diff
@@ -71,11 +71,13 @@ jobs:
               est-ra-1 --fullName "EST RA 1" --password Secret.est
           docker exec ca pki -n caadmin ca-group-member-add "EST RA Agents" est-ra-1
-          
+          docker exec ca pki -n caadmin ca-group-member-add "Certificate Manager Agents" est-ra-1
```
The check is in method `isAgentCert()` invoked in `RAClientAuthSubjectNameContraint:100` and implemented in the same class.
